### DISCUSSION TITLE
Toggle Class

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#6286] [Ideas]
I have encountered many times to toggle a class for div element. For example in creating overlays, lightbox effects, accordion effects, slider, ...

In JavaScript this is quite easy, 

See for example: https://www.w3schools.com/howto/tryit.asp?filename=tryhow_js_toggle_class

I am looking for a simple way in wikitext in Tiddlywiki to toggle classes. I did some experiments but the solution is not simple and easy to understand as I expect.

**My solution**

- use a temp tiddler
- use a toggle macro to toggle the text field in temp tiddler using `$action-listops`
- use a filtered transclusion `{{{...}}}`
- remove extra double brackets

... _too many steps in wikitext_

```
\define toggleState()
<$action-listops $tiddler=<<stateTid>> $field="text" $subfilter="+[toggle[overlay overlay-scale open],[overlay overlay-scale]]" />
\end


\define myoverlay(img, id:"1", width:"256px")
<$vars stateTid="""xx$:/temp/$(currentTiddler)$/$id$""" lb="[[" rb="]]" >
<$button class="tc-btn-invisible" actions=<<toggleState>> >
Do it
</$button>

<div class={{{ [<stateTid>get[text]trim<lb>trim<rb>] }}} >
<$button class="closebtn tc-btn-invisible" actions=<<toggleState>> > &times;</$button>
<$image source=<<__img__>> />
</div>

</$vars>
```

See how I toggle a class! I have to remove the double brackets manually! or use `$wikify`!

**Suggestion**:

- add a toggleclass to core?
- add feature to do what JS does, e.g. get the div element by id and then toggle class.
- improve the current toggle operator to work with $action-setfield
- an alternative solution using current wikitext

**Comment by saqimtiaz** [comment]
It is difficult to provide precise guidance without understanding exactly what you are trying to do, but here are a few pointers:

> \define toggleState()
<$action-listops $tiddler=<<stateTid>> $field="text" $subfilter="+[toggle[overlay overlay-scale open],[overlay overlay-scale]]" />
\end

You are passing multiple values to `toggle[]` in a single parameter which is resulting in them being wrapped in square brackets. Manipulating a list with `toggle[]` you would rewrite the above as:

```
\define toggleState()
<$action-listops $tiddler=<<stateTid>> $field="text" $subfilter="+[toggle[open]]" />
\end
```

> See how I toggle a class! I have to remove the double brackets manually! or use $wikify!

You only have square brackets due to improper use of the toggle operator. Do note however that you could also just use `enlist-input[]` to remove the brackets. However, you should not have square brackets saved in the first place for class names.

**Comment by kookma** [comment]
@saqimtiaz - thank you!
Actually I am experimenting the lightbox effects, where I need some animation and also overlay trick! I cannot use $reveal as it kills most of css transitions, animations, etc.

So, I decided to use the same approach in html5-css-js where a class is added/removed to make animations, etc.

In brief, I thought to implement the same approach in Tiddlywiki (e.g add classes / remove classes)! I was hopeful to do this simpler!

**Comment by pmario** [comment]
>

**Comment by kookma** [comment]
@saqimtiaz 

This is my second try based on your suggestion. This a demo to mimic what I said

- add/remove a class to text field of a temp tiddler
- read it by filter transclusion 

```
\define stateTid()  mystate

\define actions() 
<$list filter="[<stateTid>get[text]is[blank]]">
<$action-setfield $tiddler=<<stateTid>> $field="text" $value="overlay overlay-scale" />
</$list>
\end


<$button actions=<<actions>> > Toggle
<$action-listops $tiddler=<<stateTid>> $field="text" $subfilter="+[toggle[open]]" />
</$button>

;Result
:{{{[<stateTid>get[text]] :else[[overlay overlay-scale]] }}}
```

**NOTE**
Here I used a trick: the sequence of execution in `$button` widgets: actions from `actions` parameter, and `actions` inside `$button` widget!

**Reply by pmario** [reply]
>Here I used a trick: the sequence of execution in $button widgets: actions from actions parameter, and actions inside $button widget!

You shouldn't do this. It's hard to understand for others, who don't know the internals from TW. ... 

Your class values are also scattered all over the code. So it's hard to see, what actually should be assigned to the class. I think my example is easier to understand

**Comment by pmario** [comment]
>.. In JavaScript this is quite easy,

It's easy, because it's a minimal showcase and can't be compared with what you want to achieve. The JS example doesn't know about a state, that will survive a tiddler refresh.  So it will fail, if you open close a tiddler that contains the example. ... Hence an example that does the same things will be much more complicated. ... 

try this 

```
\define toggleState()
<$action-listops $tiddler=<<stateTid>> $field="text" $subfilter="+[toggle[open],[close]]" />
\end


\define myoverlay(img, id:"1", width:"256px")
<$vars stateTid="""xx$:/temp/$(currentTiddler)$/$id$"""  >
<$button class="tc-btn-invisible" actions=<<toggleState>> >
Do it
</$button>

<div class={{{ [<stateTid>get[text]match[open]then[overlay overlay-scale open]else[overlay overlay-scale]] }}} >
<$button class="closebtn tc-btn-invisible" actions=<<toggleState>> > &times;</$button>
<$image source=<<__img__>> />
</div>

</$vars>
\end

<<myoverlay asdf >>
```

**Comment by kookma** [comment]
@pmario thank you! 
Nice trick: you used then/else and seems more elegant compared to my solution!

**Comment by kookma** [comment]
Thank you both @saqimtiaz and @pmario 

**Conclusion**:
To toggle class of a html tag

- use a state tiddler
- use `$action-listops` with single value like `open/close` or `show/hide` to toggle the text field of state tiddler
- use filtered transclusion with `then/else` to select the right classes based on the state tiddler text value

**Comment by pmario** [comment]
That's the simplest code that I could think of at the moment. 

 - It removes the ID, which will only cause problems anyway
 - It removes the vars-wrapper
 - It makes the state a \define stateTid()

I think only the logic is there, which seems to be OK and relatively easy to understand for others.

```
\define stateTid() xx$:/temp/$(currentTiddler)$/$(__img__)$

\define toggleState()
<$action-listops $tiddler=<<stateTid>> $field="text" $subfilter="+[toggle[open],[close]]" />
\end

\define myoverlay(img, width:"256px")
<$button class="tc-btn-invisible" actions=<<toggleState>> >
Do it
</$button>

<div class={{{ [<stateTid>get[text]match[open]then[overlay overlay-scale open]else[overlay overlay-scale]] }}} >
<$button class="closebtn tc-btn-invisible" actions=<<toggleState>> > &times;</$button>
<$image source=<<__img__>> />
</div>
\end

<<myoverlay asdf >>
```
